include/aws/sdkutils/endpoints/private/ruleset_types_impl.h

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_SDKUTILS_ENDPOINTS_RULESET_TYPES_IMPL_H
+#define AWS_SDKUTILS_ENDPOINTS_RULESET_TYPES_IMPL_H
+
+#include <aws/common/ref_count.h>
+#include <aws/sdkutils/endpoints/ruleset.h>
+
+enum aws_endpoints_rule_type { AWS_ENDPOINTS_RULE_ENDPOINT, AWS_ENDPOINTS_RULE_ERROR, AWS_ENDPOINTS_RULE_TREE };
+
+enum aws_endpoints_url_type { AWS_ENDPOINTS_URL_TEMPLATE, AWS_ENDPOINTS_URL_REFERENCE, AWS_ENDPOINTS_URL_FUNCTION };
+
+enum aws_endpoints_error_type {
+    AWS_ENDPOINTS_ERROR_TEMPLATE,
+    AWS_ENDPOINTS_ERROR_REFERENCE,
+    AWS_ENDPOINTS_ERROR_FUNCTION
+};
+
+enum aws_endpoints_expr_type {
+    AWS_ENDPOINTS_EXPR_STRING,
+    AWS_ENDPOINTS_EXPR_NUMBER,
+    AWS_ENDPOINTS_EXPR_BOOLEAN,
+    AWS_ENDPOINTS_EXPR_ARRAY,
+    AWS_ENDPOINTS_EXPR_REFERENCE,
+    AWS_ENDPOINTS_EXPR_FUNCTION
+};
+
+struct aws_endpoints_parameter {
+    struct aws_allocator *allocator;
+
+    enum aws_endpoints_parameter_value_type type;
+    struct aws_string *built_in;
+
+    union {
+        struct aws_string *string;
+        bool boolean;
+    } default_value;
+
+    bool is_required;
+    struct aws_string *documentation;
+    bool is_deprecated;
+    struct aws_string *deprecated_message;
+    struct aws_string *deprecated_since;
+};
+
+struct aws_endpoints_ruleset {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+
+    struct aws_array_list rules;
+
+    struct aws_string *version;
+    struct aws_string *service_id;
+    struct aws_hash_table *parameters;
+};
+
+struct aws_endpoints_reference {
+    struct aws_allocator *allocator;
+    struct aws_string *ref;
+};
+
+struct aws_endpoints_fn;
+
+struct aws_endpoints_expr {
+    struct aws_allocator *allocator;
+    enum aws_endpoints_expr_type type;
+    union {
+        struct aws_string *string;
+        int32_t number;
+        bool boolean;
+        struct aws_array_list array;
+        struct aws_endpoints_reference *reference;
```

Comment:
why not just `struct aws_string *reference;`?

Is there a strong reason to make `aws_endpoints_reference` its own type? Does this introduce some kind of type-safety to the API? Or might we add more members to this struct in the future?

If we keep it as a struct, maybe just store it by-value like:
```c
union {
    ...
    struct aws_endpoints_reference reference;
    ...
}
```
since it's a private-and-not-reference-counted thing and its lifetime is already tied to the `aws_endpoints_expr`? you wouldn't even need to store `aws_endpoints_reference.allocator` member, since its lifetime is 100% tied to the owning `struct aws_endpoints_expr`. You could even just declare the type inline with the declaration of this union like

```c
union {
    struct aws_string *string;
    int32_t number;
    bool boolean;

    struct aws_endpoints_reference {
        struct aws_string *ref;
    } reference;
}
```